x/aol/keeper/msg_server_record.go

```diff
@@ -0,0 +1,51 @@
+package keeper
+
+import (
+	"context"
+
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/medibloc/panacea-core/x/aol/types"
+)
+
+func (k msgServer) AddRecord(goCtx context.Context, msg *types.MsgAddRecord) (*types.MsgAddRecordResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	ownerAddr, err := sdk.AccAddressFromBech32(msg.OwnerAddress)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid owner address: %v", err)
+	}
+	writerAddr, err := sdk.AccAddressFromBech32(msg.WriterAddress)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid writer address: %v", err)
+	}
+
+	topicKey := types.TopicCompositeKey{OwnerAddress: ownerAddr, TopicName: msg.TopicName}
+	if !k.HasTopic(ctx, topicKey) {
+		return nil, sdkerrors.Wrapf(types.ErrTopicNotFound, "topic <%s, %s>", msg.OwnerAddress, msg.TopicName)
+	}
+	writerKey := types.WriterCompositeKey{OwnerAddress: ownerAddr, TopicName: msg.TopicName, WriterAddress: writerAddr}
+	if !k.HasWriter(ctx, writerKey) {
+		return nil, sdkerrors.Wrapf(types.ErrWriterNotAuthorized, "writer <%s, %s, %s>", msg.OwnerAddress, msg.TopicName, msg.WriterAddress)
+	}
+
+	topic := k.GetTopic(ctx, topicKey)
+	offset := topic.NextRecordOffset()
```

Comment:
I have 1 questions.

If a another node two record request, I think sequence is duplicated.

1. `node-1` : request add record (topic : id, record : jang)
2. `node-2` : request add record (topic : id, record : lee)
3. `node-1` : get topic (topic : id, totalRecord : 1)
4. `node-2` : get topic (topic : id, totalRecord : 1)
5. `node-1` : add record (topic : id, record : jang, offset : 2)
6. `node-2` : add record (topic : id, record : lee, offset : 2)
-- create block --

and get record `offset : 2` result two records.

Umm.. There's a potential risk factor. Is not problem?

Reply:
Yeah. That's the key of the Tendermint Consensus.

All state transitions are executed sequentially, and the sequence of state transitions is synced across all nodes. That's what Tendermint guarantees.

As you mentioned at `1` and `2`, each node can have a different sequence of transactions in their [mempool](https://docs.tendermint.com/master/tendermint-core/mempool.html) like below:
|node-1|node-2|
|---|---|
|`add-record("jang")`|`add-record("yj")`|
|`add-record("yj")`|`add-record("jang")`|
It can happen frequently in distributed systems, because some clients send requests to the node-1 and others send requests to the node-2. 

However, the final sequence is determined by a block proposer.
If the node-2 is a block proposer in the `Round X`, the node-2 assemble a new block as he want (according to his mempool).
```
- Block X
    - tx1: add-record("yj")
    - tx2: add-record("jang")
```
This block is broadcasted to the node-1. Then, the node-1 plays transactions in the order specified in the block. Here, the node-1 truncates his mempool (His mempool is not meaningful anymore because he isn't a block proposer).

By this consensus mechanism, all nodes will execute all state transitions in the same order.

Reply:
I understand. Thanks